## Problem 2 - Data manipulation and selection (4 points)

In this problem you will clean the data from our data file by removing no-data values, convert temperature values in Fahrenheit to Celsius, and split the data into separate datasets using the weather station identification code. We will start this problem by cleaning and converting our temperature data. Please perform the tasks below by writing your code into the codeblocks in each section.

### Tips for completing this problem

- Use **exactly** the same variable names as in the instructions because your answers will be automatically graded, and the tests that grade your answers rely on following the same formatting or variable naming as in the instructions.
- **Please do not**:

    - **Change the file names**. Do all of your editing in the provided `Exercise-5-problem-2.ipynb` file (this file).
    - **Copy/paste cells in this notebook**. We use an automated grading system that will fail if there are copies of code cells.
    - **Change the existing cell types**. You can add cells, but changing the cell types for existing cells (from code to markdown, for example) will also cause the automated grader to fail. 

### Scores for this problem

**Your score on this problem will be based on following criteria:**

- Creating a new dataframe called `selected` that contains select columns from the data file
- Cleaning the new dataframe by removing no-data values
- Creating a new column for temperatures converted from Fahrenheit to Celsius
- Dividing the data into separate dataframes for the Helsinki Kumpula and Rovaniemi stations
- Saving the new dataframes to CSV files
- Including comments that explain what most lines in the code do
- Answering a couple questions at the end of the problem
- Uploading your notebook and data files to your GitHub repository for this week's exercise

### Part 1 (0 points)

The first step for this problem is to read the data file `6153237444115dat.csv` again into a variable `data` using pandas. Remember to specify the no-data values (you can copy your code from Problem 1).

In [13]:
data = None

# YOUR CODE HERE
import pandas as pd
data = pd.read_csv("6153237444115dat.csv" , na_values=['*', '**', '***', '****', '*****', '******'])
raise NotImplementedError()

NotImplementedError: 

Check that the first rows of the DataFrame look ok: 

In [14]:
print(data)

        USAF   WBAN  YR--MODAHRMN    DIR   SPD   GUS   CLG  SKC   L   M  ...  \
0      28450  99999  201705010000  174.0  10.0  14.0   NaN  NaN NaN NaN  ...   
1      28450  99999  201705010020  180.0  10.0   NaN   4.0  NaN NaN NaN  ...   
2      28450  99999  201705010050  190.0  10.0   NaN   4.0  NaN NaN NaN  ...   
3      28450  99999  201705010100  188.0  12.0  16.0   NaN  NaN NaN NaN  ...   
4      28450  99999  201705010120  200.0  13.0   NaN   2.0  OBS NaN NaN  ...   
...      ...    ...           ...    ...   ...   ...   ...  ...  ..  ..  ...   
11689  29980  99999  201708311900  144.0  12.0  20.0  50.0  BKN NaN NaN  ...   
11690  29980  99999  201708312000  137.0  14.0  20.0  40.0  BKN NaN NaN  ...   
11691  29980  99999  201708312100  160.0  13.0   NaN   NaN  BKN NaN NaN  ...   
11692  29980  99999  201708312200  170.0  12.0  21.0  28.0  OVC NaN NaN  ...   
11693  29980  99999  201708312300  192.0  13.0  21.0  19.0  BKN NaN NaN  ...   

          SLP    ALT     STP  MAX  MIN 

Check the number of rows in the DataFrame:

In [15]:
len(data)

11694

### Part 2 (1 point)

Next, your task is to subset the data and remove rows with missing temperature values.

- Select the columns `USAF`, `YR--MODAHRMN`, `TEMP`, `MAX`, and `MIN` from the `data` dataframe and assign them to the variable `selected`
- Remove all rows from `selected` that have NoData in the column `TEMP` using the `dropna()` function

In [16]:
selected = None
selected = data[["USAF","YR--MODAHRMN","TEMP","MAX","MIN"]]
selected_clean = selected.dropna(subset=["TEMP"])
# YOUR CODE HERE
raise NotImplementedError()

NotImplementedError: 

Check that you selected the correct column names:

In [17]:
selected.head()

USAF  YR--MODAHRMN  TEMP  MAX  MIN
0  28450  201705010000  31.0  NaN  NaN
1  28450  201705010020  30.0  NaN  NaN
2  28450  201705010050  30.0  NaN  NaN
3  28450  201705010100  31.0  NaN  NaN
4  28450  201705010120  30.0  NaN  NaN

Check how many rows you have after removing the no-data values:

In [18]:
len(selected)

11694

### Part 3 (1 point)

Next, you can convert the temperature values in Fahrenheit to Celsius.

- Create a new column in `selected` called `Celsius`
- Convert the Fahrenheit temperatures from `TEMP` using the conversion formula below and store the results in the new `Celsius` column.

$$
\Large
\begin{equation}
  T_{\mathrm{Celsius}} = (T_{\mathrm{Fahrenheit}} - 32)~/~1.8
\end{equation}
$$

- Round the values in the `Celsius` column to have 0 decimals (**do not** create a new column, update the current one)
- Convert the `Celsius` values into integers (**do not** create a new column, update the current one)

In [40]:
# YOUR CODE HERE
selected[Celsius] = 0.0
selected["Celsius"] = (selected["TEMP"]-32)/1.8
selected["Celsius"].round(0)

raise NotImplementedError()

C:\Users\IP500\AppData\Local\Temp\ipykernel_14436\1177372188.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  selected["Celsius"] = (selected["TEMP"]-32)/1.8


NotImplementedError: 

In [41]:
# Check your dataframe
selected.head()

USAF  YR--MODAHRMN  TEMP  MAX  MIN   Celsius
0  28450  201705010000  31.0  NaN  NaN -0.555556
1  28450  201705010020  30.0  NaN  NaN -1.111111
2  28450  201705010050  30.0  NaN  NaN -1.111111
3  28450  201705010100  31.0  NaN  NaN -0.555556
4  28450  201705010120  30.0  NaN  NaN -1.111111

In [35]:
# Check data types
print(selected.dtypes)

USAF              int64
YR--MODAHRMN      int64
TEMP            float64
MAX             float64
MIN             float64
Celsius         float64
dtype: object


### Part 4 (1 point)

Your next task is to divide `selected` into two separate dataframes. Please use the given variable names and write your answer to the codeblock below.

- Select all rows from the `selected` DataFrame with the `USAF` code `29980` into a variable called `kumpula`
- Select all rows from the `selected` DataFrame with the `USAF` code `28450` into a variable called `rovaniemi`

In [ ]:
kumpula = None
rovaniemi = None

# YOUR CODE HERE
raise NotImplementedError()

In [ ]:
# Check the dataframe
print(f"Kumpula: \n{kumpula.head()}\n")

In [ ]:
# Check the dataframe
print(f"Rovaniemi: \n{rovaniemi.head()}\n")

### Part 5 (1 point)

Now you can save your selections to csv files.

- Save the `kumpula` DataFrame into the file `Kumpula_temps_May_Aug_2017.csv` (CSV format) 
    - Separate the columns with commas (`,`)
    - Use only 2 decimals for the floating point numbers
- Save the `rovaniemi` DataFrame into the file `Rovaniemi_temps_May_Aug_2017.csv` (CSV format) 
    - Separate the columns with commas (`,`)
    - Use only 2 decimals for the floating point numbers
- Remember to commit both data files to your Exercise 5 repository

In [ ]:
# YOUR CODE HERE
raise NotImplementedError()

In [ ]:
#Read-only cell for hidden tests :)

### Problem 2 summary

- Was anything unclear to you in Problem 2?
- Did you encounter any problems with decimal formatting?

Also, make sure you:

- Check that your code includes informative comments explaining what your code does
- Commit and push your changes to your GitHub repository for Exercise 5 (including your 2 new data files)

Please write your answers and comments below (**activate the cell by double clicking**).

YOUR ANSWER HERE

### On to Problem 3

Now you can continue to [Problem 3: Data analysis](Exercise-5-problem-3.ipynb)